feat: Support nested .gitignore

tobni opened the pull request:
Implements a variant of solution 2 in https://github.com/pantsbuild/pants/issues/19860.

Edit: I've decided to pivot to https://github.com/pantsbuild/pants/pull/22411/commits/00c13ed245d1f6c990ae95bf5d23e7ef7b49df36 after realizing the scandir coupling causes a lot of churn.

tobni committed:
feat: Support nested .gitignore

benjyw commented:
Let us know when this is reviewable!

tobni commented on src/rust/engine/fs/src/lib.rs:
```diff
@@ -447,43 +480,48 @@ impl PosixFS {
     }
 
     pub fn is_ignored(&self, stat: &Stat) -> bool {
-        self.ignore.is_ignored(stat)
+        self.root_ignore.is_stat_ignored(stat)
```
This is only called in tests from what I can tell.

tobni commented on src/rust/engine/fs/src/lib.rs:
```diff
@@ -447,43 +480,48 @@ impl PosixFS {
-                    // It would be nice to be able to ignore paths before stat'ing them, but in order to apply
-                    // git-style ignore patterns, we need to know whether a path represents a directory.
-                    Some(Ok(s))
+                    gitignore = Some(path_to_stat);
```
Lamentations are rarely useful code comments

tobni commented:
@benjyw Ready for review!

tobni committed:
doc: Add a note

tobni commented on src/python/pants/core/goals/run_integration_test.py:
```diff
@@ -42,12 +42,12 @@ def test_restartable(use_pantsd: bool) -> None:
             ],
             workdir=workdir,
             use_pantsd=use_pantsd,
+            hermetic=False,
```
I think I need help here.

@benjyw I cannot reproduce the test failures locally, so likely it is due to cache? I am confused as to what causes the error though?

benjyw replied:
Are these happening consistently in CI? It would be great to keep this test hermetic. I can't see why your change would impact this, TBH

benjyw reviewed (commented):
Still diving into the rust side, but just some preliminary minor comments

benjyw commented on src/rust/engine/src/nodes/scandir.rs:
```diff
@@ -20,17 +21,40 @@ pub struct Scandir {
+
+    async fn parent_gitignore_stack(&self, context: &Context) -> NodeResult<GitignoreStack> {
+        let root_ignore = context.core.vfs.root_ignore();
+        if !context.core.vfs.read_gitignore_files() {
```
I'm confused: As far as I can tell, this will be true iff `use_gitignore` is true (in `Core::new()`). But if it's false shouldn't we be using no gitignores at all, not even the root one?

benjyw replied:
Ah, that is non-obvious. So how are `pants_ignore` and root `gitignore` combined?

tobni replied:
"pants_ignore", including allowlisting, is always applied first. Then git-style stack traversal follows. This is equivalent to the previous behaviour, we just previously had a stack of depth=1.

benjyw replied:
Yeah, that would make it clearer.

benjyw commented on src/rust/engine/src/nodes/scandir.rs:
```diff
@@ -20,17 +21,40 @@ pub struct Scandir {
             .core
             .vfs
-            .scandir(self.dir)
+            .scandir(self.dir, gitignore_stack)
```
This is tricky to grok, but why is the right gitignore stack to pass here the one at the parent dir of the scanned dir, and not the one at the scanned dir itself?

benjyw replied:
OK, still not sitting well in my head, but I think I get it...

benjyw reviewed (commented):
Couple more questions as I attempt to grok this code (I am not super familiar with the context)

tobni commented on docs/notes/2.29.x.md:
```diff
@@ -16,6 +16,9 @@ Thank you to [Klaviyo](https://www.klaviyo.com/) for their Platinum tier support
 ### General
 
+The `pants_ignore_use_gitignore`-option now applies to nested .gitignore files.
+This may require an update of the patterns in `pants_ignore`. In particular, after upgrading, you will have to allowlist (e.g `!pattern`) any path that you relied on pants not ignoring via nested .gitignore files.
```
I dont think we should keep the "global gitignore only" behaviour around for very long, but a feature flag might be warranted?

benjyw replied:
Yeah, I think a flag would be wise, even with a one release deprecation cycle. As long as we can detect and warn if people are relying on the old behavior?

benjyw commented on src/rust/engine/src/session.rs:
```diff
@@ -139,6 +139,12 @@ pub struct Session {
 }
 
+#[derive(Clone, Debug)]
+pub struct WeakSession {
```
This is needed because the session has a circular reference to the GitignoreSession?

benjyw commented:
Looks like some CI issues and merge conflicts. Let me know when to take a last look.

tobni committed:
fix: Do not panic when starting the InvalidationWatcher twice

A second start is now a no-op, so a repeated attach from Python cannot panic across the pyfunction boundary.

tobni committed:
fix: Start the invalidation watcher for non-pantsd runs

Core::new no longer starts the watcher, so every scheduler consumer must attach a session: add the missing attach for local runs and plugin resolution, and attach RuleRunner's real session instead of an immediately dropped one.

tobni committed:
fix: Hash .gitignore lines with a terminator

tobni committed:
feat: Compute gitignore stacks via a dedicated graph node

Approach 1 of #19860

tobni committed:
fix: Filter watch events without querying the graph when possible

tobni committed:
test: Restore hermetic run integration tests

tobni committed:
fix: Rebind the invalidation watcher to the latest attached session

Filtering no longer depends on which consumer attaches first

tobni committed:
fix: Back the bootstrap scheduler watcher with a long-lived session

cburroughs commented:
Thanks for the contribution. We've just branched for 2.33.x, so merging this pull request now will come out in 2.34.x, please move the release notes updates to docs/notes/2.34.x.md if that's appropriate.

cburroughs commented:
@tobni I see you have been pushing here again, is this ready for another round of review?
